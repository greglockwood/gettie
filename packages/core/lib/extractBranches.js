import { shouldBeIgnored } from './utils';

export default function extractBranches(data) {
  const branches = [];
  const objects = new WeakSet();

  function handleArray(array, path) {
    for (let i = 0; i < array.length; i++) {
      if (shouldBeIgnored(array, i)) {
        continue;
      }

      path.push(i);
      branches.push([...path]);
      handleItem(array[i], path);
      path.pop();
    }

    return array;
  }

  function handleObject(obj, path) {
    for (const name in obj) {
      if (shouldBeIgnored(obj, name)) {
        continue;
      }

      path.push(name);
      branches.push([...path]);
      handleItem(obj[name], path);
      path.pop();
    }

    return obj;
  }

  function handleItem(data, path) {
    if (data) {
      if (data.toJSON) {
        data = data.toJSON();
      }

      if (objects.has(data) || shouldBeIgnored(data)) {
        return data;
      }

      if (Array.isArray(data)) {
        objects.add(data);
        data = handleArray(data, path);
      } else if (data.constructor === Object) {
        objects.add(data);
        data = handleObject(data, path);
      }
    }

    return data;
  }

  handleItem(data, []);

  return branches;
}
