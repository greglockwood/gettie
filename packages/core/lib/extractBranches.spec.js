import extractBranches from './extractBranches';

describe('extract branches', () => {
  const obj = {
    a: 123,
    b: {
      [Symbol('foo')]: () => {},
      c: {
        d: 123,
      },
    },
    c: [1, { foo: 2 }, [3, 4]],
  };

  const objWithRecursion = {
    a: 123,
    b: {},
    c: [0, {}],
  };

  objWithRecursion.b.c = objWithRecursion;
  objWithRecursion.c[1] = objWithRecursion;

  it('object', () => {
    const branches = extractBranches(obj);
    expect(branches).toMatchSnapshot();
  });

  it('array', () => {
    const branches = extractBranches([obj]);
    expect(branches).toMatchSnapshot();
  });

  it('recursive', () => {
    const branches = extractBranches(objWithRecursion);
    expect(branches).toMatchSnapshot();
  });
});
