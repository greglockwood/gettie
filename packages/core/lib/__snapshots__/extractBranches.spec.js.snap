// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`extract branches array 1`] = `
Array [
  Array [
    0,
  ],
  Array [
    0,
    "a",
  ],
  Array [
    0,
    "b",
  ],
  Array [
    0,
    "b",
    "c",
  ],
  Array [
    0,
    "b",
    "c",
    "d",
  ],
  Array [
    0,
    "c",
  ],
  Array [
    0,
    "c",
    0,
  ],
  Array [
    0,
    "c",
    1,
  ],
  Array [
    0,
    "c",
    1,
    "foo",
  ],
  Array [
    0,
    "c",
    2,
  ],
  Array [
    0,
    "c",
    2,
    0,
  ],
  Array [
    0,
    "c",
    2,
    1,
  ],
]
`;

exports[`extract branches object 1`] = `
Array [
  Array [
    "a",
  ],
  Array [
    "b",
  ],
  Array [
    "b",
    "c",
  ],
  Array [
    "b",
    "c",
    "d",
  ],
  Array [
    "c",
  ],
  Array [
    "c",
    0,
  ],
  Array [
    "c",
    1,
  ],
  Array [
    "c",
    1,
    "foo",
  ],
  Array [
    "c",
    2,
  ],
  Array [
    "c",
    2,
    0,
  ],
  Array [
    "c",
    2,
    1,
  ],
]
`;

exports[`extract branches recursive 1`] = `
Array [
  Array [
    "a",
  ],
  Array [
    "b",
  ],
  Array [
    "b",
    "c",
  ],
  Array [
    "c",
  ],
  Array [
    "c",
    0,
  ],
  Array [
    "c",
    1,
  ],
]
`;
